### PR TITLE
PWGHF: Add finding leading particle in correlatorD0Hadrons

### PR DESCRIPTION
**Pull request by 1481014945**
Dear, experts
I am a newcomer.I'm sorry to bother you again. I'm Xu wang. I come from China. My email is wangxuwx@mails.ccnu.edu.cn.  
I want to start my job at PWGHF regarding the D0 meson.The subsequent tasks will also be carried out in this document.
I want to go the correlation analysis of D0 and other hadrons  with Pt>1. Then, based on the delta phi, divide it into three regions, labeled as forward, backward, and underlying. Finally, investigate the D0 mesons separately in each of these three regions.

### CODE REVIEW

**Comment by vkucera** [comment]
Hi Xu, please get familiar with the contribution guidelines:

- https://aliceo2group.github.io/analysis-framework/docs/gettingstarted/contributingtocode.html
- https://aliceo2group.github.io/analysis-framework/docs/advanced-specifics/pwghf.html

In particular, pay attention to the conventions on naming, documentation and code organisation.

If you need to change something in your PR, please do not close it, especially if there have already been comments/reviews. Creating a new PR from the same branch is just making another PR identical to the old one. You can edit the title of the PR, edit your comments and push further commits in your branch. Please do not force push which overwrites the PR commit history.

**Comment by fgrosa** [comment]
Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80. Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?

**Comment by 1481014945** [comment]
> Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> 
> https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> 
> . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?

Thank you for your reply. Sorry, I am a newcomer and I am not very clear about this process. I will try  to do it next.

**Comment by 1481014945** [comment]
> Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> 
> https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> 
> . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?

I want to search for the hadron with the highest Pt, then use that hadron as the leading particle. Afterwards, calculate delta phi and delta eta between the selected D0 and it. Based on the difference in delta phi, divide it into forward, backward, and underlying regions. Based on this information, could you provide me with some advice?

**Comment by fgrosa** [comment]
> > Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> > https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> > 
> > . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?
> 
> Thank you for your reply. Sorry, I am a newcomer and I am not very clear about this process. I will try to do it next. I want to search for the hadrons with the higher Pt, then use that hadrons as the leading particle. Afterwards, calculate delta phi and delta eta between the selected D0 and it. Based on the difference in delta phi, divide it into forward, backward, and underlying regions. Based on this information, could you provide me with some advice?

Indeed what you want to do is slightly different than what done in the task I sent you, however it would be enough in my opinion to add an option to compute correlations with respect to the leading particle instead of all the hadrons in the already existing code. However, I would like to also hear the opinion of the PAG coordinators (and @fcolamar who is more expert than me in correlations ;))

**Comment by vkucera** [comment]
Moving to draft while discussion with PAG coordinators is ongoing.

**Comment by 1481014945** [comment]
> > > > Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> > > > https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> > > > 
> > > > . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?
> > > 
> > > 
> > > Thank you for your reply. Sorry, I am a newcomer and I am not very clear about this process. I will try to do it next. I want to search for the hadrons with the higher Pt, then use that hadrons as the leading particle. Afterwards, calculate delta phi and delta eta between the selected D0 and it. Based on the difference in delta phi, divide it into forward, backward, and underlying regions. Based on this information, could you provide me with some advice?
> > 
> > 
> > Indeed what you want to do is slightly different than what done in the task I sent you, however it would be enough in my opinion to add an option to compute correlations with respect to the leading particle instead of all the hadrons in the already existing code. However, I would like to also hear the opinion of the PAG coordinators (and @fcolamar who is more expert than me in correlations ;))
> 
> Ciao @fgrosa, me and @deepathoms were discussing this topic with @1481014945 via mail. We indeed pointed him already to the existing tasks and we asked more information about the final observable he is interesting in looking at (as this was not fully clear to me, al least). We hope to have soon the opportunity to discuss deeper the analysis details!

Thank you for your reply.I think that D0hadrons and hadronsD0 are different. In lines 299 and 344 of the code above, it can be seen that D0 is looped over first, and then the tracks are looped over. However, for hadronsD0, the intention is to loop over tracks first and then loop over D0. Based on this information, could you provide me with some advice?

**Comment by fgrosa** [comment]
> > > > > Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> > > > > https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> > > > > 
> > > > > . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?
> > > > 
> > > > 
> > > > Thank you for your reply. Sorry, I am a newcomer and I am not very clear about this process. I will try to do it next. I want to search for the hadrons with the higher Pt, then use that hadrons as the leading particle. Afterwards, calculate delta phi and delta eta between the selected D0 and it. Based on the difference in delta phi, divide it into forward, backward, and underlying regions. Based on this information, could you provide me with some advice?
> > > 
> > > 
> > > Indeed what you want to do is slightly different than what done in the task I sent you, however it would be enough in my opinion to add an option to compute correlations with respect to the leading particle instead of all the hadrons in the already existing code. However, I would like to also hear the opinion of the PAG coordinators (and @fcolamar who is more expert than me in correlations ;))
> > 
> > 
> > Ciao @fgrosa, me and @deepathoms were discussing this topic with @1481014945 via mail. We indeed pointed him already to the existing tasks and we asked more information about the final observable he is interesting in looking at (as this was not fully clear to me, al least). We hope to have soon the opportunity to discuss deeper the analysis details!
> 
> Thank you for your reply.I think that D0hadrons and hadronsD0 are different. In lines 299 and 344 of the code above, it can be seen that D0 is looped over first, and then the tracks are looped over. However, for hadronsD0, the intention is to loop over tracks first and then loop over D0. Based on this information, could you provide me with some advice?

Hi @1481014945. In general I do not think that looping first on the tracks is good. In fact, if you do so you cannot check that the leading particle that you select is not a D0 daughter (and correlating a D0 with one of its daughters does not make much sense).

**Comment by 1481014945** [comment]
> > > > > > Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> > > > > > https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> > > > > > 
> > > > > > . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?
> > > > > 
> > > > > 
> > > > > Thank you for your reply. Sorry, I am a newcomer and I am not very clear about this process. I will try to do it next. I want to search for the hadrons with the higher Pt, then use that hadrons as the leading particle. Afterwards, calculate delta phi and delta eta between the selected D0 and it. Based on the difference in delta phi, divide it into forward, backward, and underlying regions. Based on this information, could you provide me with some advice?
> > > > 
> > > > 
> > > > Indeed what you want to do is slightly different than what done in the task I sent you, however it would be enough in my opinion to add an option to compute correlations with respect to the leading particle instead of all the hadrons in the already existing code. However, I would like to also hear the opinion of the PAG coordinators (and @fcolamar who is more expert than me in correlations ;))
> > > 
> > > 
> > > Ciao @fgrosa, me and @deepathoms were discussing this topic with @1481014945 via mail. We indeed pointed him already to the existing tasks and we asked more information about the final observable he is interesting in looking at (as this was not fully clear to me, al least). We hope to have soon the opportunity to discuss deeper the analysis details!
> > 
> > 
> > Thank you for your reply.I think that D0hadrons and hadronsD0 are different. In lines 299 and 344 of the code above, it can be seen that D0 is looped over first, and then the tracks are looped over. However, for hadronsD0, the intention is to loop over tracks first and then loop over D0. Based on this information, could you provide me with some advice?
> 
> Hi @1481014945. In general I do not think that looping first on the tracks is good. In fact, if you do so you cannot check that the leading particle that you select is not a D0 daughter (and correlating a D0 with one of its daughters does not make much sense).

1263e690330f60e0973da65ecc1f277e

Can't we skip the correlation between D0 daughters and D0 using a similar approach?

**Comment by fgrosa** [comment]
> > > > > > > Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> > > > > > > https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> > > > > > > 
> > > > > > > . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?
> > > > > > 
> > > > > > 
> > > > > > Thank you for your reply. Sorry, I am a newcomer and I am not very clear about this process. I will try to do it next. I want to search for the hadrons with the higher Pt, then use that hadrons as the leading particle. Afterwards, calculate delta phi and delta eta between the selected D0 and it. Based on the difference in delta phi, divide it into forward, backward, and underlying regions. Based on this information, could you provide me with some advice?
> > > > > 
> > > > > 
> > > > > Indeed what you want to do is slightly different than what done in the task I sent you, however it would be enough in my opinion to add an option to compute correlations with respect to the leading particle instead of all the hadrons in the already existing code. However, I would like to also hear the opinion of the PAG coordinators (and @fcolamar who is more expert than me in correlations ;))
> > > > 
> > > > 
> > > > Ciao @fgrosa, me and @deepathoms were discussing this topic with @1481014945 via mail. We indeed pointed him already to the existing tasks and we asked more information about the final observable he is interesting in looking at (as this was not fully clear to me, al least). We hope to have soon the opportunity to discuss deeper the analysis details!
> > > 
> > > 
> > > Thank you for your reply.I think that D0hadrons and hadronsD0 are different. In lines 299 and 344 of the code above, it can be seen that D0 is looped over first, and then the tracks are looped over. However, for hadronsD0, the intention is to loop over tracks first and then loop over D0. Based on this information, could you provide me with some advice?
> > 
> > 
> > Hi @1481014945. In general I do not think that looping first on the tracks is good. In fact, if you do so you cannot check that the leading particle that you select is not a D0 daughter (and correlating a D0 with one of its daughters does not make much sense).
> 
> 1263e690330f60e0973da65ecc1f277e
> 
> Can't we skip the correlation between D0 daughters and D0 using a similar approach?

Technically yes, but if I properly understood what you aim to do, the principle is very different. In the example you reported, we reject DD pairs when they share a daughter track because we are sure that one of the two is background.

If I understood properly instead, you want to use the leading pT particle as proxy of the jet direction and measure the yield in the forward, backward and transverse regions. So, In the case that the leading pT particle is the D0 daughter (and I guess it's in a large fraction of the events with a D0), then you propose to simply skip the D0. However this creates a bias because by construction you reject many D0 mesons in the forward region. 
I would suggest to discuss it briefly in the HFC PAG before going on with the implementation. You can maybe prepare 1-2 slides with Physics motivation and how you technically implement it to have a more efficient discussion. Thanks a lot!

**Comment by 1481014945** [comment]
> > > > > > > > Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> > > > > > > > https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> > > > > > > > 
> > > > > > > > . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?
> > > > > > > 
> > > > > > > 
> > > > > > > Thank you for your reply. Sorry, I am a newcomer and I am not very clear about this process. I will try to do it next. I want to search for the hadrons with the higher Pt, then use that hadrons as the leading particle. Afterwards, calculate delta phi and delta eta between the selected D0 and it. Based on the difference in delta phi, divide it into forward, backward, and underlying regions. Based on this information, could you provide me with some advice?
> > > > > > 
> > > > > > 
> > > > > > Indeed what you want to do is slightly different than what done in the task I sent you, however it would be enough in my opinion to add an option to compute correlations with respect to the leading particle instead of all the hadrons in the already existing code. However, I would like to also hear the opinion of the PAG coordinators (and @fcolamar who is more expert than me in correlations ;))
> > > > > 
> > > > > 
> > > > > Ciao @fgrosa, me and @deepathoms were discussing this topic with @1481014945 via mail. We indeed pointed him already to the existing tasks and we asked more information about the final observable he is interesting in looking at (as this was not fully clear to me, al least). We hope to have soon the opportunity to discuss deeper the analysis details!
> > > > 
> > > > 
> > > > Thank you for your reply.I think that D0hadrons and hadronsD0 are different. In lines 299 and 344 of the code above, it can be seen that D0 is looped over first, and then the tracks are looped over. However, for hadronsD0, the intention is to loop over tracks first and then loop over D0. Based on this information, could you provide me with some advice?
> > > 
> > > 
> > > Hi @1481014945. In general I do not think that looping first on the tracks is good. In fact, if you do so you cannot check that the leading particle that you select is not a D0 daughter (and correlating a D0 with one of its daughters does not make much sense).
> > 
> > 
> > 1263e690330f60e0973da65ecc1f277e
> > Can't we skip the correlation between D0 daughters and D0 using a similar approach?
> 
> Technically yes, but if I properly understood what you aim to do, the principle is very different. In the example you reported, we reject DD pairs when they share a daughter track because we are sure that one of the two is background.
> 
> If I understood properly instead, you want to use the leading pT particle as proxy of the jet direction and measure the yield in the forward, backward and transverse regions. So, In the case that the leading pT particle is the D0 daughter (and I guess it's in a large fraction of the events with a D0), then you propose to simply skip the D0. However this creates a bias because by construction you reject many D0 mesons in the forward region. I would suggest to discuss it briefly in the HFC PAG before going on with the implementation. You can maybe prepare 1-2 slides with Physics motivation and how you technically implement it to have a more efficient discussion. Thanks a lot!

For this issue, I consulted my supervisor, and he said this's a good point, indeed. But a leading particle from heavy flavor decay should be fine. We should not reject the D0 if it is a daughter of the leading particle as it is assumed from hard processes. Regarding the slice, I am currently attempting to work on it.

**Comment by fgrosa** [comment]
> > > > > > > > > Hi @1481014945, sorry if I did not see before this PR. I was however wondering whether what you aim to do cannot be done already with this task
> > > > > > > > > https://github.com/AliceO2Group/O2Physics/blob/3b59cec13fa2918b50d8031cda9b311bbb475d34/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx#L80
> > > > > > > > > 
> > > > > > > > > . Did you discuss with the HFC coordinators (@mmazzilli @deepathoms) about this?
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Thank you for your reply. Sorry, I am a newcomer and I am not very clear about this process. I will try to do it next. I want to search for the hadrons with the higher Pt, then use that hadrons as the leading particle. Afterwards, calculate delta phi and delta eta between the selected D0 and it. Based on the difference in delta phi, divide it into forward, backward, and underlying regions. Based on this information, could you provide me with some advice?
> > > > > > > 
> > > > > > > 
> > > > > > > Indeed what you want to do is slightly different than what done in the task I sent you, however it would be enough in my opinion to add an option to compute correlations with respect to the leading particle instead of all the hadrons in the already existing code. However, I would like to also hear the opinion of the PAG coordinators (and @fcolamar who is more expert than me in correlations ;))
> > > > > > 
> > > > > > 
> > > > > > Ciao @fgrosa, me and @deepathoms were discussing this topic with @1481014945 via mail. We indeed pointed him already to the existing tasks and we asked more information about the final observable he is interesting in looking at (as this was not fully clear to me, al least). We hope to have soon the opportunity to discuss deeper the analysis details!
> > > > > 
> > > > > 
> > > > > Thank you for your reply.I think that D0hadrons and hadronsD0 are different. In lines 299 and 344 of the code above, it can be seen that D0 is looped over first, and then the tracks are looped over. However, for hadronsD0, the intention is to loop over tracks first and then loop over D0. Based on this information, could you provide me with some advice?
> > > > 
> > > > 
> > > > Hi @1481014945. In general I do not think that looping first on the tracks is good. In fact, if you do so you cannot check that the leading particle that you select is not a D0 daughter (and correlating a D0 with one of its daughters does not make much sense).
> > > 
> > > 
> > > 1263e690330f60e0973da65ecc1f277e
> > > Can't we skip the correlation between D0 daughters and D0 using a similar approach?
> > 
> > 
> > Technically yes, but if I properly understood what you aim to do, the principle is very different. In the example you reported, we reject DD pairs when they share a daughter track because we are sure that one of the two is background.
> > If I understood properly instead, you want to use the leading pT particle as proxy of the jet direction and measure the yield in the forward, backward and transverse regions. So, In the case that the leading pT particle is the D0 daughter (and I guess it's in a large fraction of the events with a D0), then you propose to simply skip the D0. However this creates a bias because by construction you reject many D0 mesons in the forward region. I would suggest to discuss it briefly in the HFC PAG before going on with the implementation. You can maybe prepare 1-2 slides with Physics motivation and how you technically implement it to have a more efficient discussion. Thanks a lot!
> 
> For this issue, I consulted my supervisor, and he said this's a good point, indeed. But a leading particle from heavy flavor decay should be fine. We should not reject the D0 if it is a daughter of the leading particle as it is assumed from hard processes. Regarding the slice, I am currently attempting to work on it.

Hi @1481014945, without considering the concerns about the analysis (which we discussed today in HFC), for what concerns the code, I would suggest you to define a function that does [this](https://github.com/AliceO2Group/O2Physics/pull/4962/files#diff-814a19e75103805fc33a7ef4b69fa1f82d3d7d60b4df554e129417ba203b0e7eR120-R131) in the existing code (https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx) and it returns the index of the leading-pT track, which is executed only in case a configurable is set as `true`. Then, if enabled via this configurable, for each D0 candidate, you skip all the tracks that have a different index. This allows you to essentially re-use the existing code for your analysis. Let me know if this makes sense to you!

**Review by vkucera — CHANGES_REQUESTED** [review]
Your last commit e6d7ac55e7ae97523059e13f0501a81667ffd213 reverts recent fixes. Please fix that. How did you make that commit?

**Comment by vkucera** [comment]
As a general advice, please dedicated some time to learning basic Git commands and using GitHub. You can find links to some materials in the official documentation of the O2 analysis framework.

**Review by fgrosa — APPROVED** [review]

**Comment by fgrosa** [comment]
Hi @1481014945 the latest implementation looks good to me, thanks! It seems to me that also the reverted fix pointed out by @vkucera has been now corrected (but I let him cross check to be sure).

File: PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx [hunk 1]
```diff
@@ -264,6 +266,21 @@ struct HfCorrelatorD0Hadrons {
...
   }
 
+  // Find Leading Particle
+  template <typename TTracks>
+  int findLeadingParticle(TTracks const& tracks)
+  {
+    auto leadingParticle = tracks.begin();
+    for (auto const& track : tracks) {
+      if (std::abs(track.dcaXY()) >= 1. || std::abs(track.dcaZ()) >= 1.)
+        continue;
```
Comment (vkucera) [inline PWGHF/HFC/TableProducer/correlatorD0Hadrons.cxx:276]:
```suggestion
      if (std::abs(track.dcaXY()) >= 1. || std::abs(track.dcaZ()) >= 1.) {
        continue;
      }
```

**Review by vkucera — CHANGES_REQUESTED** [review]
Minor change, otherwise looks good.

**Comment by 1481014945** [comment]
Regarding the minor issue you mentioned, I have already made the changes. Are there any other issues?

**Comment by fgrosa** [comment]
Hi @vkucera, if it looks good also to you I would say that we can merge it (you just also need to approve, otherwise we cannot merge it). Thanks!

**Review by vkucera — APPROVED** [review]

**Pull request merged**